File: apps/agent/tests/properties/_template_test.py
"""Hypothesis PBT template — copy into apps/agent/tests/properties/ for a
sub-spec. Name the property (P-XXX-NN) and reference foundation-setup/design.md
§Correctness Properties.
"""
from hypothesis import given
from hypothesis import strategies as st

from agent.retry import TransientProviderError, with_retry


@given(st.integers(min_value=1, max_value=5))
def test_with_retry_attempts_bounded(max_attempts):
    """Property: with_retry calls the fn at most max_attempts times."""
    calls = {"n": 0}

    def always_transient():
        calls["n"] += 1
        raise TransientProviderError("x")

    try:
        with_retry(always_transient, max_attempts=max_attempts, sleep=lambda _: None)
    except TransientProviderError:
        pass
    assert calls["n"] == max_attempts

File: tests/properties/_template.test.ts
import { describe, it, expect } from "vitest";
import fc from "fast-check";

/**
 * PBT template — copy this into tests/properties/<your-spec>/ when starting a
 * sub-spec. State the correctness property (P-XXX-NN) in the describe title and
 * reference the foundation-setup/design.md §Correctness Properties entry.
 *
 * Convention: arbitraries up top, the invariant inside fc.assert(fc.property(...)).
 */
describe("P-TEMPLATE: example invariant", () => {
  it("array concat length is additive (sample property)", () => {
    fc.assert(
      fc.property(fc.array(fc.integer()), fc.array(fc.integer()), (a, b) => {
        expect(a.concat(b).length).toBe(a.length + b.length);
      }),
    );
  });
});
